docs/ADR-10-cardano-api-script-witness-api.md

```diff
@@ -0,0 +1,217 @@
+
+```haskell
+
+data TxScriptWitnessRequirements era
+  = TxScriptWitnessRequirements
+      (Set L.Language)
+      [L.Script era]
+      (L.TxDats era)
+      (L.Redeemers era) -- Redeemer pointer map
+
+getTxScriptWitnessesRequirements
+  :: AlonzoEraOnwards era
+  -> [(Witnessable witnessable (ShelleyLedgerEra era), AnyWitness (ShelleyLedgerEra era))]
+  -> TxScriptWitnessRequirements (ShelleyLedgerEra era)
+getTxScriptWitnessesRequirements eon wits =
+  obtainMonoidConstraint eon $ mconcat $ map (getTxScriptWitnessRequirements eon) wits
+```
+
+## New `IndexedPlutusScriptWitness` types
+
+We are only concerned with the index of something that is plutus script witnessed. Therefore we introduce `IndexedPlutusScriptWitness`:
+
+```haskell
+data IndexedPlutusScriptWitness witnessable (lang :: L.Language) (purpose :: PlutusScriptPurpose) era where
+  IndexedPlutusScriptWitness
+    :: Witnessable witnessable era
+    -> (L.PlutusPurpose L.AsIx era)
+    -> (PlutusScriptWitness lang purpose era)
+    -> IndexedPlutusScriptWitness witnessable lang purpose era
+```
+
+We do away with `cardano-api`'s `ScriptWitnessIndex` and directly use `cardano-ledger`'s `PlutusPurpose AsIx era` which is essentially the same thing. 
+The benefit here is we get to use ledger type classes to calculate the index of the plutus script witness thing we are interested in:
+
+```haskell
+class GetPlutusScriptPurpose era where
+  toPlutusScriptPurpose
+    :: Word32
+    -> Witnessable thing era
+    -> L.PlutusPurpose L.AsIx era
+
+instance GetPlutusScriptPurpose era where
+  toPlutusScriptPurpose index WitTxIn{} = L.mkSpendingPurpose (L.AsIx index)
+  toPlutusScriptPurpose index WitWithdrawal{} = L.mkRewardingPurpose (L.AsIx index)
+  toPlutusScriptPurpose index WitMint{} = L.mkMintingPurpose (L.AsIx index)
+  toPlutusScriptPurpose index WitTxCert{} = L.mkCertifyingPurpose (L.AsIx index)
+  toPlutusScriptPurpose index WitVote{} = L.mkVotingPurpose (L.AsIx index)
+  toPlutusScriptPurpose index WitProposal{} = L.mkProposingPurpose (L.AsIx index)
+```
+
+This ultimately allows us to collapse all of the different index calculation functions into a single function:
+
+```haskell
+createIndexedPlutusScriptWitness
+  :: Word32
+  -> Witnessable witnessable era
+  -> PlutusScriptWitness lang purpose era
+  -> IndexedPlutusScriptWitness witnessable lang purpose era
+createIndexedPlutusScriptWitness index witnessable =
+  IndexedPlutusScriptWitness witnessable (toPlutusScriptPurpose index witnessable)
+
+createIndexedPlutusScriptWitnesses
+  :: [(Witnessable witnessable era, AnyWitness era)]
+  -> [AnyIndexedPlutusScriptWitness era]
+createIndexedPlutusScriptWitnesses witnessableThings =
+  [ AnyIndexedPlutusScriptWitness $ createIndexedPlutusScriptWitness index thing sWit
+  | (index, (thing, AnyPlutusScriptWitness sWit)) <- zip [0 ..] witnessableThings
```

Comment:
- `TxIn` is indexed on sorted `TxIn`s.
- `TxWithdrawals` are indexed on the order of stake address
- `TxVotingProcedures` are indexed on the order of votes
- `TxProposalProcedures` are indexed on their order

Here you are proposing indexing everything by just the order in the list (which is not sorted!) which does not look right to me.

Reply:
Ah I forgot to update the ADR. Added the ordering function.